apollo-router/src/spec/schema.rs

```diff
@@ -22,68 +25,81 @@ use crate::Configuration;
-        api_schema: Option<Box<Schema>>,
-    ) -> Result<Self, SchemaError> {
-        let mut schema = Self::parse_inner(s, configuration)?;
-        schema.api_schema = api_schema;
-        Ok(schema)
-    }
-
     #[cfg(test)]
     pub(crate) fn parse_test(s: &str, configuration: &Configuration) -> Result<Self, SchemaError> {
-        let mut schema = Self::parse_inner(s, configuration)?;
-        schema.api_schema = Some(Box::new(Self::parse_inner(
-            &make_api_schema(s)?,
-            configuration,
-        )?));
+        let api_schema = Self::parse(&make_api_schema(s, configuration)?, configuration)?;
+        let schema = Self::parse(s, configuration)?.with_api_schema(api_schema);
         Ok(schema)
     }
 
-    fn parse_inner(schema: &str, _configuration: &Configuration) -> Result<Schema, SchemaError> {
+    pub(crate) fn parse(sdl: &str, configuration: &Configuration) -> Result<Self, SchemaError> {
         let mut compiler = ApolloCompiler::new();
-        let id = compiler.add_type_system(schema, "schema.graphql");
+        let id = compiler.add_type_system(sdl, "schema.graphql");
 
         let ast = compiler.db.ast(id);
 
         // Trace log recursion limit data
         let recursion_limit = ast.recursion_limit();
         tracing::trace!(?recursion_limit, "recursion limit data");
 
-        // TODO: run full compiler-based validation instead?
-        let errors = ast.errors().cloned().collect::<Vec<_>>();
-        if !errors.is_empty() {
-            let errors = ParseErrors {
-                raw_schema: schema.to_string(),
-                errors,
+        let diagnostics =
+            if configuration.experimental_graphql_validation == GraphQLValidation::Legacy {
+                vec![]
+            } else {
+                compiler
+                    .validate()
+                    .into_iter()
+                    .filter(|err| err.data.is_error())
+                    .collect::<Vec<_>>()
+            };
+
+        // Only bail out on parser errors for now: validation errors will be checked with the query
+        // planner result
+        if diagnostics
+            .iter()
+            .any(|err| matches!(*err.data, DiagnosticData::SyntaxError { .. }))
+        {
+            let errors = ValidationErrors {
+                errors: diagnostics,
             };
             errors.print();
             return Err(SchemaError::Parse(errors));
         }
 
-        fn as_string(value: &hir::Value) -> Option<&String> {
-            if let hir::Value::String { value, .. } = value {
-                Some(value)
-            } else {
-                None
+        if !diagnostics.is_empty() {
+            let errors = ValidationErrors {
+                errors: diagnostics.clone(),
+            };
+            errors.print();
+
+            if configuration.experimental_graphql_validation == GraphQLValidation::New {
+                return Err(SchemaError::Parse(errors));
             }
```

Comment:
Shouldn’t this branch also be taken when config is "both"? Or are stored diagnostics used later? Please add a code comment to explain.